pkg/wharfapi/problem.go

```diff
@@ -1,61 +1,51 @@
-// defined by IETF RFC-7807.
-type Problem struct {
-	// Type is a URI reference that identifies the problem type. The IETF
-	// RFC-7807 specification encourages that, when dereferenced, it provide
-	// human-readable documentation for the problem type (e.g., using HTML).
-	// When this member is not present, its value is assumed to be
-	// "about:blank".
-	Type string `json:"type" example:"https://iver-wharf.github.io/#/prob/build/run/invalid-input"`
-
-	// Title is a short, human-readable summary of the problem type.
-	// It SHOULD NOT change from occurrence to ocurrence of the problem, except
-	// for purposes of localization.
-	//
-	// Recommended to be kept brief, have proper punctuation, and be
-	// capitalized, like a short sentence.
-	Title string `json:"title" example:"Invalid input variable for build."`
-
-	// Status is the HTTP status code generated by the origin server for this
-	// occurrence of the problem.
-	Status int `json:"status" example:"400"`
 
-	// Detail is a human-readable explanation specific to this occurrence of the
-	// problem.
-	//
-	// Recommended to have proper punctuation, and be capitalized,
-	// like a sentence. Compared to Title this field may stretch on and be
-	// longer.
-	Detail string `json:"detail" example:"Build requires input variable 'myInput' to be of type 'string', but got 'int' instead."`
-
-	// Instance is a URI reference that identifies the specific occurrence of
-	// the problem. It may or may not yield further information if dereferenced.
-	Instance string `json:"instance" example:"/projects/12345/builds/run/6789"`
+	"github.com/iver-wharf/wharf-core/pkg/problem"
+)
 
-	// Error is an extended field for the regular Problem model defined in
-	// RFC-7807. It contains the string message of the error (if any).
-	Errors []string `json:"errors" example:"strconv.ParseUint: parsing \"-1\": invalid syntax"`
-}
+const problemContentType = "application/problem+json"
 
-func (r Problem) Error() error {
+func newProblemError(r problem.Response) error {
 	trimmedTitle := strings.TrimRight(firstRuneLower(r.Title), ",.!; ")
 	return ProblemError{
 		Message: fmt.Sprintf("%s: %s", trimmedTitle, strings.Join(r.Errors, "; ")),
 		Problem: r,
 	}
 }
 
+func isProblemResponse(r *http.Response) bool {
+	return r.Header.Get("Content-Type") == problemContentType
+}
+
+func parseProblemResponse(response *http.Response) (problem.Response, error) {
+	resp, readErr := ioutil.ReadAll(response.Body)
+	if readErr != nil {
+		return problem.Response{}, fmt.Errorf(
+			"failed to read problem response body: %w", readErr)
+	}
+	if closeErr := response.Body.Close(); closeErr != nil {
+		return problem.Response{}, fmt.Errorf(
+			"failed to close problem response body reading: %w", closeErr)
+	}
+	var prob problem.Response
+	if jsonErr := json.Unmarshal(resp, &prob); jsonErr != nil {
+		return problem.Response{}, fmt.Errorf(
+			`failed to parse "%s" problem response: %w`, problemContentType, jsonErr)
+	}
+	return prob, nil
+}
```

Comment:
Maybe some of these could be moved to core?

Reply:
Good idea!

Reply:
https://github.com/iver-wharf/wharf-core/pull/12

Reply:
Will change this repo in a new PR once that wharf-core one has been merged.